overlord/managers_test.go

```diff
@@ -1955,6 +1956,348 @@ type: kernel`
+
+	ts, _, err := snapstate.InstallPath(st, &snap.SideInfo{RealName: "pc-kernel"}, kernelSnapPath, "", "", snapstate.Flags{})
+	c.Assert(err, IsNil)
+	chg := st.NewChange("install-snap", "...")
+	chg.AddAll(ts)
+
+	// run, this will trigger a wait for the restart
+	st.Unlock()
+	err = s.o.Settle(settleTimeout)
+	st.Lock()
+	c.Assert(err, IsNil)
+
+	c.Assert(bloader.BootVars, DeepEquals, map[string]string{
+		"kernel_status": boot.TryStatus,
+	})
+
+	// we are in restarting state and the change is not done yet
+	restarting, _ := st.Restarting()
+	c.Check(restarting, Equals, true)
+	c.Check(chg.Status(), Equals, state.DoingStatus)
+
+	// the kernelSnapInfo we mocked earlier will not have a revision set for the
+	// SideInfo, but since the previous revision was "1", the next revision will
+	// be x1 since it's unasserted, so we can set the Revision on the SideInfo
+	// here to make comparison easier
+	kernelSnapInfo.SideInfo.Revision = snap.R(-1)
+
+	// check that we extracted the kernel snap assets
+	extractedKernels := bloader.ExtractKernelAssetsCalls
+	c.Assert(extractedKernels, HasLen, 1)
+	c.Assert(extractedKernels[0].Filename(), Equals, kernelSnapInfo.Filename())
+
+	// and that we enabled the try kernel
+	enabledTryKernels, _ := bloader.GetRunKernelImageFunctionSnapCalls("EnableTryKernel")
+	c.Assert(enabledTryKernels, HasLen, 1)
+	c.Assert(enabledTryKernels[0].Filename(), Equals, kernelSnapInfo.Filename())
+
+	// we won't disable any try kernels nor will we enable any kernels
+	_, nDisableTryKernelCalls := bloader.GetRunKernelImageFunctionSnapCalls("DisableTryKernel")
+	c.Assert(nDisableTryKernelCalls, Equals, 0)
+	_, nEnableKernelCalls := bloader.GetRunKernelImageFunctionSnapCalls("EnableKernel")
+	c.Assert(nEnableKernelCalls, Equals, 0)
+
+	// pretend we restarted
+	s.mockSuccessfulReboot(c, bloader)
+
+	st.Unlock()
+	err = s.o.Settle(settleTimeout)
+	st.Lock()
+	c.Assert(err, IsNil)
+
+	// we should now have enabled a kernel
+	enabledKernels, _ := bloader.GetRunKernelImageFunctionSnapCalls("EnableKernel")
+	c.Assert(enabledKernels, HasLen, 1)
+	c.Assert(enabledKernels[0].Filename(), Equals, kernelSnapInfo.Filename())
+
+	// we should have now disabled a TryKernel
+	_, nDisableTryKernelCalls = bloader.GetRunKernelImageFunctionSnapCalls("DisableTryKernel")
+	c.Assert(nDisableTryKernelCalls, Equals, 1)
+
+	c.Assert(chg.Status(), Equals, state.DoneStatus, Commentf("install-snap change failed with: %v", chg.Err()))
+
+	// also check that we are active on the second revision
+	var snapst snapstate.SnapState
+	err = snapstate.Get(st, "pc-kernel", &snapst)
+	c.Assert(err, IsNil)
+	c.Check(snapst.Sequence, HasLen, 2)
+	c.Check(snapst.Sequence, DeepEquals, []*snap.SideInfo{si1, &kernelSnapInfo.SideInfo})
+	c.Check(snapst.Active, Equals, true)
+	c.Check(snapst.Current, DeepEquals, snap.R(-1))
+}
+
+func (s *mgrsSuite) TestInstallKernelSnap20UndoUpdatesBootloaderEnv(c *C) {
+	bloader := bootloadertest.Mock("mock", c.MkDir()).UC20RunModeRebootReady()
```

Comment:
Bit annoying that there's only a handfull of lines changes in this test compared to the successful path one. However, this is a quite common in our codebase. Wonder if there's some sort of optimization, other than splitting this into a helper function, we could use here that would allow sharing the setup and provide enough flexibility to execute corresponding verification code that won't otherwise feel awkward to use.

Reply:
I go back and forth on this a lot. OTOH, yes less duplicated code is good because it eliminates issues like copy-pasting unnecessary or even wrong bits for tests, but OTOOH, being less clever in tests and more explicit about what we're setting up I think makes the tests more readable and easier to understand - there's less magic to them which sometimes is what we want.